dotnet/src/Connectors/Connectors.Amazon/Bedrock/Core/Clients/BedrockTextGenerationClient.cs

```diff
@@ -0,0 +1,160 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.Text.Json;
+using System.Text.Json.Nodes;
+using Amazon.BedrockRuntime;
+using Amazon.BedrockRuntime.Model;
+using Connectors.Amazon.Core.Requests;
+using Connectors.Amazon.Core.Responses;
+using Connectors.Amazon.Models;
+using Connectors.Amazon.Models.AI21;
+using Connectors.Amazon.Models.Amazon;
+using Connectors.Amazon.Models.Anthropic;
+using Connectors.Amazon.Models.Cohere;
+using Connectors.Amazon.Models.Meta;
+using Connectors.Amazon.Models.Mistral;
+
+namespace Microsoft.SemanticKernel.Connectors.Amazon.Core;
+
+/// <summary>
+/// Represents a client for interacting with the text generation through Bedrock.
+/// </summary>
+/// <typeparam name="TRequest"> Request object which is an ITextGenerationRequest. </typeparam>
+/// <typeparam name="TResponse"> Response object which is an ITextGenerationResponse. </typeparam>
+public abstract class BedrockTextGenerationClient<TRequest, TResponse>
+    where TRequest : ITextGenerationRequest
+    where TResponse : ITextGenerationResponse
+{
+    private readonly string _modelId;
+    private readonly IAmazonBedrockRuntime _bedrockApi;
+    private readonly IBedrockModelIOService<ITextGenerationRequest, ITextGenerationResponse> _ioService;
+
+    /// <summary>
+    /// Builds the client object and registers the model input-output service given the user's passed in model ID.
+    /// </summary>
+    /// <param name="modelId"></param>
+    /// <param name="bedrockApi"></param>
+    /// <exception cref="ArgumentException"></exception>
+    protected BedrockTextGenerationClient(string modelId, IAmazonBedrockRuntime bedrockApi)
+    {
+        this._modelId = modelId;
+        this._bedrockApi = bedrockApi;
+        string[] parts = modelId.Split('.'); //modelId looks like "amazon.titan-text-premier-v1:0"
+        string modelProvider = parts[0];
+        string modelName = parts.Length > 1 ? parts[1] : string.Empty;
+        switch (modelProvider)
```

Comment:
Is this a duplicate to what is in the `BedrockChatCompletionClient` constructor? If so this should be refactored into a common method. Especially since this is what will need to be updated everytime Bedrock adds a new model.

Reply:
Sounds good. I've created a BedrockClientIOService class which serves as the common utility class for extracting the IO service and model provider. Now new services can just call those functions and new models can just be added there